src/extensions/ExtensionBase.js

```diff
@@ -0,0 +1,34 @@
+import { isValidExtensionType } from './utilities.js';
+
+export class ExtensionBase {
+
+	constructor( extType, extName ) {
+
+		// the key used to look up into the extension objects
+		this.name = extName;
+
+		this.extensionType = extType;
+
+		// enforce a 3d-tiles spec supported extensions blob, which we know how to extract
+		if ( ! isValidExtensionType( extType ) ) throw Error( `Unsupported Extension Type: ${extType}` );
+
+		this.data = {};
+
+	}
+
+	useTileset( tileset ) {
+
+		// if the extension is used after a root tileset has been loaded
+		// this should have any root level data associated with the extension.name
+		// ex: for the `3DTILES_content_gltf` extension, it should include the extensionsUsed/Required to load the gltf files
+		// referenced in the tileset.
+		// - "3DTILES_content_gltf": { "extensionsUsed": ["EXT_mesh_gpu_instancing"], "extensionsRequired": ["EXT_mesh_gpu_instancing"] }
+		// see extension specs @ https://github.com/CesiumGS/3d-tiles/tree/main/extensions for what you might find in the .data object.
+
+		// initialize the base data from the top level for this extension if available
+		this.data = tileset && tileset.extensions && tileset.extensions[ this.name ] || {};
```

Comment:
I think the biggest questionmark is how the extension gets access to the various extension data blobs.

the GLTFExtension is *very much* the trivial case that doesn't need*any* of this extra extension functionality.

Here's a slightly contrived, very untested example custom vendor extension that is closer to the sort of thing I wanted to do in the holiday hackery which spawned this PR. 

Namely: Split a single file downloaded over the wire into separate tiles in the tileset, but keeping a valid  per-spec tileset when extension isn't available. (Split .cmpt-ish, but instead of multiple contents within a single tile all sharing the same bounding volume, child tiles have more constrained bounds).

```typescript
import { ExtensionBase, ExtensionType, GLTFExtensionLoader } from '3d-tiles-renderer';
import { DefaultLoadingManager, LoadingManager } from 'three';
import { GLTF } from 'three/examples/jsm/loaders/GLTFLoader';

interface ExtensionTileData {
  sourceUri: string;
  sceneId: number;
}

interface ExtensionTilesetData {
  gltfs: { [key: string]: Promise<GLTF>|undefined };
  requests: { [key: string]: Promise<Response>|undefined };
}

export class SceneSelectionExtension extends ExtensionBase {
  public data: ExtensionTilesetData;
  constructor(private manager: LoadingManager = DefaultLoadingManager) {
    super( ExtensionType.CONTENT, 'MTTR_extension_scenetile' );
  }

  // *if* we already have a GLTF parsed/loaded to select a scene from, skip the fetch.
  // if not, request source scene, and track as pending
  public fetch(extDataTile: ExtensionTileData|null, url: string, ...args: any[]): null|Promise<any> {

    // Top level tileset extension data:
    // extDataTile => tile.content.extensions.MTTR_extension_scenetile: { sourceUri: '0.glb', sceneId: 'room4' }
    const sourceUri = extDataTile?.sourceUri;
    const sceneId = extDataTile?.sceneId;
    if (!sourceUri || !sceneId) return null;

    // tile specific extension data:
    // extDataTileset tileset.extensions.MTTR_extension_scenetile.gltfs: { [key: sourceUri]: Promise<GLTF> }
    const extDataTileset = this.data;
    const gltfLoadPromise = extDataTileset.gltfs[sourceUri];
    let request = extDataTileset.requests[sourceUri];

    // if a fully parsed/loaded gltf to select from in parse already exists, skip the fetch altogether
    if (gltfLoadPromise) {
      return Promise.resolve({ ok: true, arrayBuffer: () => Promise.resolve(null) });
    }
    // TODO: handle multiple fetches so that we don't parse the full gltf 1x
    // for each time it is requested while downloading if this looks like
    // a halfway valid approach (heh).
    else if (request) {
      return request;
    }
    // otherwise, download the source file to extract from, and return that request
    else {
      // TODO: sourceUri needs same url signing content.uri/url has already had done earlier.
      request = fetch(sourceUri, ...args);
      // flag as download in progress
      extDataTileset.requests[sourceUri] = request;
      return request;
    }
  }

  // Respond with the three object3d directly from cached data, no need to run through gltf parsing again.
  public parse(extDataTile: ExtensionTileData, buffer: ArrayBuffer|null, fileExt: string, workingPath: string, fetchOptions: RequestInit): null|Promise<any> {

    const sceneId = extDataTile?.sceneId;
    const sourceUri = extDataTile?.sourceUri;
    if (!sceneId || !sourceUri) return null;

    // tileset.extensions.MTTR_extension_scenetile.gltfs: { [key: sourceUri]: Promise<GLTF> }
    const extDataTileset = this.data;

    // if we have a real buffer here, then that means it needs to be
    let gltfPromise = extDataTileset.gltfs[sourceUri];
    if (!gltfPromise) {
      if (!buffer) {
        throw Error(`TODO: MTTR_extension_scenetile => no buffer to parse!`);
      }
      // store *full* GLTF, not just the single main scene
      const loader = new GLTFExtensionLoader(this.manager);
      loader.workingPath = workingPath;
      loader.fetchOptions = fetchOptions;
      gltfPromise = loader.parse(buffer);
      extDataTileset.gltfs[sourceUri] = gltfPromise;
    } else {
      // and select the specific scene/nodes/etc the extension wants to use
      // for that tile
      const selectedScenePromise = gltfPromise.then(gltf => {
        const selectedScene = gltf.scenes[sceneId];
        if (!selectedScene) {
          throw Error(`TODO: MTTR_extension_scenetile => scene ${sceneId} not found in ${sourceUri}`);
        }
        return selectedScene;
      });
      return selectedScenePromise;
    }

    return null;
  }
}
```

One problem I'm having is that I have **some** hope to keep the tilesets themselves at least minimally usable (if not fully optimized for perf?) by other 3d-tiles systems without extension or application work being REQUIRED. With this extension as optional, the tileset is still totally viewable, just kinda slow for excessively large scenes with thousands of tiles. While with extension available the player can be nice and responsive, and load much less content over the wire.

Reply:
Sorry I'm having a hard to understanding what exactly you want to do. Is this a good interpretation?

- Have a valid tile set with a tile that references a GLTF file with many sub objects.
- At run time load that GLTF with many sub objects and create a series of "dummy" child tiles of that parent tile to stand in for the subobjects. Those child tiles will have bounding volumes that tightly encompass the sub objects they represent. The parent tile is now considered "empty".
- When one of these dummy tiles is set to be loaded / displayed you want to stuff in one of those sub objects instead of fetching and parsing a new file

Is that correct? Would this only occur for leaf tiles? Which part of the tile update / render process is slow that you're trying to render here?

Reply:
That's the goal of that extension yes, though after giving it a little more thought over the weekend, we have some other alternatives to look into also before going with something this extreme.

>Would this only occur for leaf tiles? Which part of the tile update / render process is slow that you're trying to optimize here?

The solve is more about working around a deficiency in the spatial subdivision choices, and adapting existing data for use with 3d-tiles than it is an issue of tile update / render.

Reply:
> The solve is more about working around a deficiency in the spatial subdivision choices, and adapting existing data for use with 3d-tiles than it is an issue of tile update / render.

I'd have to understand a bit more about the spatial subdivision limitations you're running in to. Maybe it's worth creating an issue to talk about different performance limitations with tilesets (spec and implementation driven) to see what we can come up with.

Maybe it's tangential at this point but regarding performance there are a few ideas I'd had (but haven't needed in my work) that could help with geometry and rendering updates:

- http2 should allow for faster parallel downloads.
- A plugin could be created for retrieving geometry which could fall back to fetch but otherwise defer to IndexedDB storage to forgo a full download and parse of a file every time it's needed.
- The recent "BatchedMesh" proposal for three.js could be used in limited scenarios to improve draw calls across the full tileset.